### PR TITLE
Build transformed distributions within compiled code

### PR DESCRIPTION
**Pull request by vmoens**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #135001

cc @fritzo @neerajprad @alicanb @nikitaved @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

**Comment by vmoens** [comment]
This PR tests that we can build an execute a distribution within a compiled code without graph break.
Currently, all tests first build the dist / transform - I add a couple of tests for compile in these cases - but in practice people will likely build their dist on the fly. In this case, `weakref.ref` in `inv` causes issues.

I assume we have a `weakref` there to cache the result of the `inv` without risking having floating references in the code. This can be removed in compiled code where the advantage of caching the inv is less clear. We still need the `ComposeTransform` to keep a ref to `self` within `inv` though, as people may rely on the fact that `t.inv.inv is t` is `True`.

**Review by anijain2305 — COMMENTED** [review]
We should be able to support this in Dynamo, I think.

**Comment by vmoens** [comment]
> We should be able to support this in Dynamo, I think.
> 
> 

I guess so, but as of now if you comment out my changes and run the new test it'll fail. Note that the error message is also a bit obscure when that happens within torchrl: it mentions a call to `None.__new__` that is unsupported...

**Comment by vmoens** [comment]
@anijain2305 this is still broken on nightlies:
```python
from torch.distributions import TransformedDistribution, Normal, ComposeTransform, AffineTransform
import torch

def make_dist_and_test():
    dist = TransformedDistribution(
        Normal(0, 1), ComposeTransform([AffineTransform(0, 1)])
    )
    s = dist.rsample((1,))
    return s, dist.log_prob(s)

make_dist_and_test = torch.compile(make_dist_and_test, fullgraph=True)
make_dist_and_test()
```
results in 
```
torch._dynamo.exc.Unsupported: Graph break due to unsupported builtin None.ReferenceType.__new__. This function is either a Python builtin (e.g. _warnings.warn) or a third-party C/C++ Python extension (perhaps created with pybind). If it is a Python builtin, please file an issue on GitHub so the PyTorch team can add support for it and see the next case for a workaround. If it is a third-party C/C++ Python extension, please either wrap it into a PyTorch-understood custom operator (see https://pytorch.org/tutorials/advanced/custom_ops_landing_page.html for more details) or, if it is traceable, use torch.compiler.allow_in_graph.
```

Happy to help implement a more orthodox solution though!

**Comment by vmoens** [comment]
@anijain2305 

Unfortunately this isn't still fixed.

I think (like it is the case in #138080), the issue here is that a code that runs when the transform is created outside of the compiled region and queried within runs fine, but if the transform is created within the compiled region, access to a weakref that points to it triggers an exception.

**Comment by anijain2305** [comment]
I will take a look tomorrow.

File: torch/distributions/transforms.py [hunk 1]
```diff
@@ -125,7 +125,8 @@ def inv(self):
             inv = self._inv()
         if inv is None:
             inv = _InverseTransform(self)
-            self._inv = weakref.ref(inv)
+            if not torch.compiler.is_dynamo_compiling():
+                self._inv = weakref.ref(inv)
```
Comment (fritzo) [inline torch/distributions/transforms.py:129]: If dynamo not support `weakref`s, I would guess the workaround would be to replace each `weakref` with some sort of api compatible `strongref` and just accept that the resulting reference cycles will need a gc pass to free memory, rather than `weakref` which needs only refcounts.

Comment (vmoens) [reply torch/distributions/transforms.py:129]: I think compile should support weakrefs but I wonder if the issue isn't that they're supported as long as the object has been created outside of the compiled region, not within.
